[ENH] Implement a SimpleCausalModel class

Aniketsy opened the pull request:
Fixes #2426 

Please let me know if my approach or fix needs any improvements . I’m open to feedback and happy to make changes based on suggestions.
Thankyou !

Aniketsy committed:
[ENH] Move the orientations argument in ExpertInLoop into ExpertKnowledge

Aniketsy committed:
[ENH] Move the orientations argument in ExpertInLoop into ExpertKnowledge

Aniketsy committed:
[ENH] Move the orientations argument in ExpertInLoop into ExpertKnowledge

Aniketsy committed:
[ENH] Move the orientations argument in ExpertInLoop into ExpertKnowledge

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

ankurankan commented on pgmpy/base/DAG.py:
Would suggest putting this class in a separate file.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1925,3 +1925,46 @@ def __hash__(self):
+
+class SimpleCausalModel(DAG):
+    """
+    A specialized DAG class for simple causal models.
```
The docstring needs to be expanded quite a bit. It should also have Parameters and Examples sections. Refer other classes for examples.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1925,3 +1925,46 @@ def __hash__(self):
+    """
+
+    def __init__(
+        self, exposure, outcome, adjustment=None, instrument=None, latents=None
```
`covariates` seems like a more appropriate name instead of `adjustment`.

ankurankan replied:
We should also add a `mediators` argument here. If mediators are present, the edges would be: `exposure -> mediator` and `mediator -> outcome`

ankurankan replied:
Would suggest using plural term for each of them as we are allowing multiple variables for each.

ankurankan replied:
Can we add type hinting here?

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1925,3 +1925,46 @@ def __hash__(self):
+            if isinstance(var, str):
+                return [var]
+            elif isinstance(var, int):
+                return [f"Var_{var}"]
```
This behavior should be explained in the docstring.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1925,3 +1925,46 @@ def __hash__(self):
+            elif var is None:
+                return []
+            return list(var)
+
```
We should not define functions inside methods unless absolutely necessary.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1925,3 +1925,46 @@ def __hash__(self):
+        # Add edges from instrument variables to exposure variables
+        for inst in self.instrument:
+            for exp in self.exposure:
+                self.add_edge(inst, exp)
```
In addition to adding the edges, we need to define the variable roles as well. Have a look at the `__init__` method of the `DAG` class for examples. I would suggest restructuring the method to first construct the list of edges that would be present in the DAG, and then, at the end, do a `super` call using the constructed edge list and variable roles.

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -2216,3 +2216,73 @@ def test_sebastiani_round_trip_comparison(self):
             )
+
+
+class TestSimpleCausalModel(unittest.TestCase):
```
Please put this in a new test file. Also, try to use pytest style tests, as we are trying to move from unit tests and are slowly refactoring the tests.

ankurankan replied:
The tests will also need to check that the roles for the variables are correctly set.

ankurankan reviewed (changes requested):
@Aniketsy Thanks for the PR. I have added a few comments, please have a look.

Aniketsy commented:
@ankurankan Thanks for the review. I will update the PR as per your suggestions.

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy commented:
@ankurankan  could you please review these changes? I’m getting one test failure and would appreciate your guidance.
Also, I’m facing some difficulty running the tests locally. Could you please guide me on the correct setup or point me to the relevant documentation?

ankurankan commented on pgmpy/base/simple_causal_model.py:
```diff
@@ -0,0 +1,140 @@
+    This class simplifies the creation of causal graphs commonly used in causal inference,
+    where the structure consists of exposures, outcomes, covariates (confounders), mediators, and instruments.
+    It automatically adds the standard edges:
+        - Exposures -> Outcomes
```
This is not true; this edge is only added if there are no mediators. 

Is it possible to show an ASCII figure here showing the graph that would be created?

ankurankan commented on pgmpy/base/simple_causal_model.py:
```diff
@@ -0,0 +1,140 @@
+    Parameters
+    ----------
+    exposures: str, int, or iterable
+        The exposure variable(s). If an int is provided, the variable name will be generated as "Var_<int>".
```
Would suggest encoding the variable names depending on their types. Exposures can be "E_<int>", outcomes can be "O_<int>", covariates can be "X_<int>", and so on. It would make it easier to identify what is the role of each node.

ankurankan commented on pgmpy/base/simple_causal_model.py:
```diff
@@ -0,0 +1,140 @@
+    -----
+    If any of the `exposures`, `outcomes`, `covariates`, `mediators`, or `instruments` arguments are
+    provided as integers, they will be automatically converted to variable names in the format
+    "Var_<int>".
```
Not required; already explained above.

ankurankan commented on pgmpy/base/simple_causal_model.py:
```diff
@@ -0,0 +1,140 @@
+
+    Examples
+    --------
+    >>> from pgmpy.base.simple_causal_model import SimpleCausalModel
```
Add `SimpleCausalModel` to `pgmpy.base.__init__` and simplify the import path.

ankurankan commented on pgmpy/base/simple_causal_model.py:
```diff
@@ -0,0 +1,140 @@
+    ... )
+    >>> model.edges()
+    OutEdgeView([('X', 'Y'), ('Z', 'X'), ('Z', 'Y'), ('I', 'X'), ('X', 'M'), ('M', 'Y')])
+        If any of the `exposures`, `outcomes`, `covariates`, `mediators`, or `instruments` arguments are
```
Any text inside code blocks requires one blank line above and below. Refer to numpydoc style documentation for details: https://numpydoc.readthedocs.io/en/latest/format.html

ankurankan commented on pgmpy/base/simple_causal_model.py:
```diff
@@ -0,0 +1,140 @@
+        edges = []
+
+        # Add edges from exposures to outcomes
+        edges += [(exp, out) for exp in exposures for out in outcomes]
```
This should only be done if there are no mediators. If there are mediators, it should be exposure -> mediator -> outcome.

ankurankan commented on pgmpy/base/simple_causal_model.py:
```diff
@@ -0,0 +1,140 @@
+        edges += [(exp, med) for exp in exposures for med in mediators]
+        edges += [(med, out) for med in mediators for out in outcomes]
+
+        super().__init__(edges)
```
You can pass latents and variable roles here itself. Have a look at the signature of `DAG.__init__`.

ankurankan commented on pgmpy/base/simple_causal_model.py:
```diff
@@ -0,0 +1,140 @@
+        set
+            A set of variable names associated with the role.
+        """
+        return self.variable_roles.get(role, set())
```
This class inherits `DAG` class which inherits `_GraphRolesMixin`. Since `_GraphRolesMixin` already has `get_role` method, it should automatically work with this class as well.

ankurankan commented on pgmpy/base/SimpleCausalModel.py:
```diff
@@ -0,0 +1,140 @@
+            return []
+        elif isinstance(var, str):
+            return [var]
+        elif isinstance(var, int):
```
I don't think the behavior here is correct. When an integer is passed, we should create that many variables with that role.

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy commented:
Thanks for the detailed review, @ankurankan , I hope I'm not creating extra work for you. I’ve applied the changes as per your suggestions. Please let me know if anything needs further improvement.

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

ankurankan commented on pgmpy/base/simple_causal_model.py:
```diff
@@ -0,0 +1,163 @@
+        - Mediators -> Outcomes
+
+    Diagram
+    -------
```
I am not sure but I don't think we can have `Diagram` section. I don't know how Sphinx would parse it.

ankurankan commented on pgmpy/base/simple_causal_model.py:
```diff
@@ -0,0 +1,163 @@
+        Exposures ---> Mediators
+           |              |
+           v              v
+        Outcomes <--------
```
This doesn't look correct: Exposures to outcomes edges should not be present. A more standard way to draw these models is to place the instrument, exposure, mediator, and outcome in a single line.

ankurankan commented on pgmpy/base/SimpleCausalModel.py:
```diff
@@ -0,0 +1,163 @@
+        - Covariates -> Outcomes
+        - Instruments -> Exposures
+        - Exposures -> Mediators
+        - Mediators -> Outcomes
```
I would suggest adding a line here saying that if you want to have more control on the model structure, use the `DAG` class.

ankurankan commented on pgmpy/base/simple_causal_model.py:
```diff
@@ -0,0 +1,163 @@
+    Parameters
+    ----------
+    exposures: str, int, or iterable
+        The exposure variable(s). If an int ``n`` is provided, ``n`` variables
```
Shouldn't the `n` be within single quote? What does using double quotes do?

ankurankan commented on pgmpy/base/simple_causal_model.py:
```diff
@@ -0,0 +1,163 @@
+                "covariates": "X_",
+                "mediators": "M_",
+                "instruments": "I_",
+            }.get(role, "Var_")
```
If the variables are an iterable (i.e., the names of the variables are provided), we shouldn't add prefix.

ankurankan commented on pgmpy/base/simple_causal_model.py:
```diff
@@ -0,0 +1,163 @@
+
+        if latents_set:
+            self.add_nodes_from(latents_set)
+            self.latents = latents_set
```
Why is this needed? Aren't the latents getting added in line 158?

Aniketsy replied:
I added this because the DAG class requires all latents to be present as nodes, otherwise, it raises a `ValueError` during initialization. please correct me if i'm wrong.

ankurankan commented on pgmpy/tests/test_base/test_simple_causal_model.py:
```diff
@@ -0,0 +1,124 @@
+
+def test_integer_variables():
+    model = SimpleCausalModel(
+        exposures=1, outcomes=2, covariates=3, mediators=4, instruments=5
```
Can you add some latents here?

ankurankan reviewed (changes requested):
Mostly looks good now. I have added a few minor comments.

kajal-jotwani committed:
Update `_GraphRolesMixin` to support multiple roles per node (#2431)

### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this
repository.

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix
branch** (right side). Don't request your master!
- [ ] Make sure you are making a pull request against the **dev branch**
(left side). Also you should start *your branch* off *our dev*.
- [ ] Check the commit's or even all commits' message styles matches our
requested structure.

### Issue number(s) that this pull request fixes
- Fixes #2427 

### List of changes to the codebase in this pull request
- Refactored _GraphRolesMixin to store roles as a set ("roles") rather
than a single string ("role") so nodes can hold multiple roles at once.
- Updated with_role to append the new role to the existing set instead
of overwriting it.
- Updated get_role, get_roles, and without_role methods to handle the
set structure.
- Updated tests in test_AncestralBase.py to fix failures and verify that
multiple roles are persisted correctly.

kajal-jotwani committed:
Fix doctests for NaiveAdjustmentRegressor (#2441)

### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this
repository.

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix
branch** (right side). Don't request your master!
- [ ] Make sure you are making a pull request against the **dev branch**
(left side). Also you should start *your branch* off *our dev*.
- [ ] Check the commit's or even all commits' message styles matches our
requested structure.

### Issue number(s) that this pull request fixes
- Fixes #2413 

### List of changes to the codebase in this pull request
- before:
<img width="500" height="300" alt="image"
src="https://github.com/user-attachments/assets/7a6646b3-7ab2-40fb-880a-e756eefc7f22"
/>

- after: 
<img width="500" height="300" alt="image"
src="https://github.com/user-attachments/assets/45b4c23e-1c10-4904-8afc-7060a5a2b795"
/>

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy committed:
[ENH] Implement a SimpleCausalModel class

Aniketsy commented:
@ankurankan  Thanks for the review! I've updated the diagram section as you mentioned. Please take a look and let me know if anything else needs improvement. all checks are green now.

ankurankan committed:
Merge branch 'fix-2426' of github.com:Aniketsy/pgmpy into pr/2434

ankurankan committed:
Modifies behaviour to throw an error if the specified latent doesn't exist in the graph

ankurankan commented:
@Aniketsy Looks good to me. I have added a commit with the following changes:
1. Renames the files as we use CamelCase for file names.
2. Adds the SimpleCausalGraph to `base/__init__.py` file to simplify import paths.
3. Modifies the behaviour of how latents are handled. Usually, existing nodes in the graph are specified as latent variables. So, the change now throws an error if the latent variable doesn't already exist in the graph instead of adding them.

Aniketsy commented:
@ankurankan Thanks! for the changes and improvements.
